mcstatus/motd.py

```diff
@@ -0,0 +1,820 @@
+}
+
+MINECRAFT_COLOR_TO_ANSI_8_BIT_JAVA = MINECRAFT_COLOR_TO_ANSI_8_BIT_BEDROCK.copy()
+MINECRAFT_COLOR_TO_ANSI_8_BIT_JAVA[MinecraftColor.GOLD] = ("214", "234")
+
+
+@dataclass
+class WebColor:
+    """Raw HTML color from MOTD.
+
+    Can be found in MOTD when someone uses gradient.
+    """
+
+    hex: str
+    rgb: tuple[int, int, int]
+
+    @classmethod
+    def from_hex(cls, hex: str) -> Self:
+        """Parse hex color, from RGB.
+
+        :raises ValueError: When hex color isn't correct, so we can't generate RGB for it.
+        :returns: New ``WebColor`` instance.
+        """
+        hex = hex.lstrip("#")
+
+        try:
+            rgb = tuple(int(hex[i : i + 2], 16) for i in (0, 2, 4))
+        except ValueError:
+            raise ValueError(f"Failed to parse given hex color: {'#' + hex!r}")
+
+        return cls.from_rgb(rgb)
+
+    @classmethod
+    def from_rgb(cls, rgb: tuple[int, int, int]) -> Self:
+        """Parse RGB color, from hex.
+
+        :raises ValueError: When RGB color is out of its 8-bit range.
+        :returns: New ``WebColor`` instance.
+        """
+        if not all(255 >= e >= 0 for e in rgb):
+            raise ValueError("Got RGB color out of its 8-bit range (0-255)")
+
+        hex = "#{:02x}{:02x}{:02x}".format(*rgb)
+        return cls(hex, rgb)
+
+
+@dataclass
+class TranslationTag:
+    """Represents a ``translate`` field in server's answer.
+
+    This just exist, and completely ignored by our end-parsers (``ansi``, ``html`` etc. properties).
+    You can parse them with ``Motd.parsed`` attribute.
+
+    .. seealso:: `Minecraft's wiki. <https://minecraft.fandom.com/wiki/Raw_JSON_text_format#Translated_Text>`_
+    """
+
+    id: str
+
+
+ParsedMotdItem: TypeAlias = "Formatting | MinecraftColor | WebColor | TranslationTag | str"
+
+
+@dataclass
+class Motd:
+    """Represents parsed MOTD.
+
+    Basing on this class, you can write your own MOTD-to-something parser, see ``parsed`` field.
+
+    :param parsed: Parsed MOTD, which then will be transformed.
+        Bases on this field, you can easily write your own MOTD-to-something parser.
+    :param bedrock: Is server Bedrock Edition? Some details may change in work of this class.
+    """
+
+    parsed: list[ParsedMotdItem]
+    bedrock: bool = False
+
+    # Replace those if you want custom transform logic
+    HTML_TRANSFORMER: type[HtmlTransformer] | None = field(repr=False, hash=False, compare=False, default=None)
+    ANSI_3_BIT_TRANSFORMER: type[Ansi3BitTransformer] | None = field(repr=False, hash=False, compare=False, default=None)
+    ANSI_4_BIT_TRANSFORMER: type[Ansi4BitTransformer] | None = field(repr=False, hash=False, compare=False, default=None)
+    ANSI_8_BIT_TRANSFORMER: type[Ansi8BitTransformer] | None = field(repr=False, hash=False, compare=False, default=None)
+    ANSI_24_BIT_TRANSFORMER: type[Ansi24BitTransformer] | None = field(repr=False, hash=False, compare=False, default=None)
```

Comment:
> I don't really like this kind of approach. If this is meant as a way for the user to "easily" override these and add their custom approaches for transforming the motd, why do this in such a convoluted way when python already has a proper way of doing it, being to make a child class and just override that specific property.

Well, yeah. Maybe user should use
```py
>>> MyTransformer(status.motd).transform()
...
>>> # Instead of
>>> status.motd.html_transformer = MyTransformer
>>> status.motd.html
...
```

> These transformers are hardly something that requires classes, they could just as easily be simple functions.

User should have a possibility to modify only concert steps in class logic, and as functions they were too complex.

> If we want to avoid repetition, why not do it sensibly and just create a private function like:

I do not think that users will modify only tables for transforming (though they also should be possible to modify). As example, `TranslationTag`.

> Also, this apparent "ease of overriding" with this approach is not even necessary here. After all, these based on some pretty well defined standards, so as long as they're working correctly on the library side, changing how they work shouldn't be needed, and we shouldn't advise it.

They work correctly, but user would like to modify many kinds of behavior. Like `TranslationTag` or `Formatting.OBFUSCATED` in HTML transformer. These things are not in this library scope, and will hardly depend on the user needs.

> I doubt that most people will end up needing more than the minecraft and plaintext versions. The ansi versions are nice to have though, if someone wants colored text in terminal, and the html ones can obviously be quite useful too, however there isn't really that much else you could need. So, if almost no one will actually need this, why even have support for it?

I also thought about it, people most likely will not use it. But I would leave at least HTML transformer, as it could be helpful in much more scenarios than ANSI ones.

> - Having a `transform` method on `Transformer` class, that takes in a `Motd` instance and returns a string.

It's already in - `BaseTransformer.__init__` does it.

> However this approach would be much more annoying for the user, than just accessing it via a property. For that reason, I don't really think it's justified, especially since again, I don't think most people will actually have the need to create their own transformers, so my first suggestion holds.

It will. For this, exist properties, that are only aliases to actual transformers. Maybe, for custom transformers, user shouldn't modify anything in `Motd` instance, and instead - pass it to custom transformer class.

Reply:
Just to note, mcstatus.io and mcsrvstat.us provide only `raw` (ours `minecraft`), `clean` (ours `plain`) and `html` variants of MOTD.

Reply:
I re-read my pre-previous comment, and looks like I wrote something strange... I will re-write it soon.

Reply:
Okay, I re-read it. My proposal is to remove `ANSI` transformers entirely, and left only `plain`, `minecraft` and `html`. And remove base class for transformers. Only `html` transformer will be class, as it's really complex. `plain` and `minecraft` will be functions.

I added it to my To-Do, and if you agree - set 👍, please.

Reply:
As I said, I think we should be consistent and there's absolutely no reason to use a class for `html` transformations, if it's too complex, split it into multiple functions, no need for a class.

Also, ANSI is pretty helpful and it could end up being used quite often in terminal applications, I don't think we should drop support for it.

Reply:
> if it's too complex, split it into multiple functions

I really dislike the idea, of having multiple function without explicitly marked goal (like class/module/package). Will be okay `transformers.py` from your package splitting?

Reply:
I don't mind the functions living in `transformers.py`, but you shouldn't use classes purely to be basically single-use function wrappers, that's not what classes are for.

If you want classes, they need to be reusable, just like in my example, where the class instances basically represent the transformer configuration, and you can use that configuration to perform multiple transformations with the transform method.

Reply:
Fixed [here](https://github.com/py-mine/mcstatus/compare/a8a1a5b59a663740f5c91918bf88ac6aa1cfec27..e53b5dd85b9e267c18d8532707b02e8181b6337b).